File: Data Structures & Algorithms/add-two-numbers/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        newList = ListNode()
        curr = newList
        carry = 0
        while l1 or l2:
            n1 = l1.val if l1 else 0
            n2 = l2.val if l2 else 0
            new = carry + n1 + n2
            val = new % 10
            carry = new // 10
            curr.next = ListNode(val = val)
            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        curr.next = ListNode(val = carry) if carry else None
        return newList.next

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) {return nullptr;}
        while (lists.size() > 1) {
            vector<ListNode*> merged;
            for (int i = 0; i < lists.size(); i += 2) {
                ListNode* next_node = i + 1 < lists.size() ? lists[i + 1] : nullptr;
                ListNode* new_node = merge(lists[i], next_node);
                merged.push_back(new_node);
            }
            lists = merged;
        } 
        return lists[0];
    }
    ListNode* merge(ListNode* node1, ListNode* node2) {
        if (!node1) {return node2;}
        if (!node2) {return node1;}
        ListNode dummy;
        ListNode* curr = &dummy;
        while (node1 and node2) {
            if (node1->val < node2->val) {
                curr->next = node1;
                node1 = node1->next;
            } else {
                curr->next = node2;
                node2 = node2->next;
            }
            curr = curr->next;
        }
        curr->next = node1 ? node1 : node2;
        return dummy.next;
    }
};
